File: Call-Android/app/src/main/java/org/ar/call/MainActivity.java
package org.ar.call;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.gyf.immersionbar.ImmersionBar;
import com.kongzue.dialog.v3.MessageDialog;
import com.kongzue.dialog.v3.TipDialog;
import com.kongzue.dialog.v3.WaitDialog;
import com.yanzhenjie.permission.AndPermission;
import com.yanzhenjie.permission.runtime.Permission;

import org.ar.call.multiuser.MultiUserActivity;
import org.ar.call.p2p.CallActivity;
import org.ar.call.p2p.VideoActivity;
import org.ar.call.utils.NetCheckUtil;
import org.ar.rtm.ErrorInfo;
import org.ar.rtm.LocalInvitation;
import org.ar.rtm.RemoteInvitation;
import org.ar.rtm.ResultCallback;
import org.ar.rtm.RtmCallEventListener;
import org.ar.rtm.RtmClientListener;
import org.ar.rtm.RtmMessage;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Map;

public class MainActivity extends BaseActivity{

    private TextView tvUser;
    private String userId;
    private Button btnLogin;
    private LinearLayout llCallType;
    private boolean isLoginSuccess;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ImmersionBar.with(this).statusBarDarkFont(false,0.2f).init();
        btnLogin = findViewById(R.id.btn_login);
        llCallType = findViewById(R.id.ll_call_type);

        tvUser = findViewById(R.id.tv_user);
        userId = CallApplication.the().getUserId();
        tvUser.setText("您的呼叫ID:"+userId);
        if (!AndPermission.hasPermissions(this,Permission.RECORD_AUDIO,Permission.CAMERA,Permission.WRITE_EXTERNAL_STORAGE,Permission.READ_EXTERNAL_STORAGE)){
            AndPermission.with(this)
                    .runtime()
                    .permission(Permission.READ_EXTERNAL_STORAGE,Permission.WRITE_EXTERNAL_STORAGE,Permission.CAMERA,Permission.RECORD_AUDIO)
                    .onGranted(permissions -> {
                        // Storage permission are allowed.
                    })
                    .onDenied(permissions -> {
                        // Storage permission are not allowed.
                    })
                    .start();
        }
        Login(false);
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }

    public void Login(boolean needTip) {
        CallApplication.the().getCallManager().getRtmClient().login("", userId, new ResultCallback<Void>() {
            @Override
            public void onSuccess(Void var1) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        isLoginSuccess =true;
                        if (needTip){
                            TipDialog.show(MainActivity.this, "登录成功！", TipDialog.TYPE.SUCCESS);
                        }
                    }
                });
            }

            @Override
            public void onFailure(final ErrorInfo var1) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        isLoginSuccess = false;
                        if (needTip) {
                            TipDialog.show(MainActivity.this, "登录失败！", TipDialog.TYPE.ERROR);
                        }
                    }
                });
            }
        });

    }


    public void showExitDialog() {
        MessageDialog.show(this, "提示", "确定要退出吗？", "确定")
                .setCancelButton("取消", (baseDialog, v) ->{
                    baseDialog.doDismiss();
                    return true;
                } )
                .setOnOkButtonClickListener((baseDialog, v) -> {
                    baseDialog.doDismiss();
                    CallApplication.the().ReleaseAll();
                    System.exit(0);
                    finish();
                    return true;
                });
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode==KeyEvent.KEYCODE_BACK){
            showExitDialog();
           return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    public void P2P(View view) {

        if (isLoginSuccess){
            if (!AndPermission.hasPermissions(this,Permission.CAMERA,Permission.RECORD_AUDIO)){
                Toast.makeText(this,"请开启相机和录音权限",Toast.LENGTH_SHORT).show();
                return;
            }
            Intent intent = new Intent(MainActivity.this, CallActivity.class);
            startActivity(intent);
        }else {
            Login(true);
        }


    }

    public void Multiple(View view) {
        if (isLoginSuccess) {
            if (!AndPermission.hasPermissions(this,Permission.CAMERA,Permission.RECORD_AUDIO)){
                Toast.makeText(this,"请开启相机和录音权限",Toast.LENGTH_SHORT).show();
                return;
            }
            Intent intent = new Intent(MainActivity.this, MultiUserActivity.class);
            startActivity(intent);
        }else {
            Login(true);
        }
    }

    @Override
    public void onRemoteInvitationReceived(RemoteInvitation remoteInvitation) {
        super.onRemoteInvitationReceived(remoteInvitation);
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                try {
                    JSONObject jsonObject = new JSONObject(remoteInvitation.getContent());
                    boolean isConference = jsonObject.getBoolean("Conference");
                    if (!isConference){
                        Intent i = new Intent(MainActivity.this, VideoActivity.class);
                        i.putExtra("RecCall",true);
                        startActivity(i);
                    }else {
                        Intent i = new Intent(MainActivity.this, MultiUserActivity.class);
                        i.putExtra("RecCall",true);
                        startActivity(i);
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }

            }
        });
    }
}

File: Call-Android/app/src/main/java/org/ar/call/MainActivity.kt
package org.ar.call

import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import com.gyf.immersionbar.ImmersionBar
import com.kongzue.dialog.interfaces.OnDialogButtonClickListener
import com.kongzue.dialog.util.BaseDialog
import com.kongzue.dialog.v3.MessageDialog
import com.kongzue.dialog.v3.TipDialog
import com.lzf.easyfloat.EasyFloat
import com.yanzhenjie.permission.AndPermission
import com.yanzhenjie.permission.runtime.Permission
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.ar.call.CallApp.Companion.callApp
import org.ar.call.databinding.ActivityMainBinding
import org.ar.call.multi.MultiCallActivity
import org.ar.call.p2p.CallActivity
import org.ar.call.p2p.VideoActivity
import org.ar.rtm.ErrorInfo
import org.ar.rtm.RemoteInvitation
import org.ar.rtm.ResultCallback
import org.json.JSONObject
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine
import kotlin.system.exitProcess


class MainActivity : BaseActivity() {

    private lateinit var binding: ActivityMainBinding
    private var isLoginSuccess = false
    private var loginTipDialog: MessageDialog? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        ImmersionBar.with(this).statusBarDarkFont(false, 0.2f).init()
        binding.tvUser.text="您的呼叫ID:${callApp.userId}"

        AndPermission.with(this)
                    .runtime()
                    .permission(Permission.WRITE_EXTERNAL_STORAGE, Permission.CAMERA, Permission.RECORD_AUDIO)
                    .onGranted { _: List<String?>? ->
                        lifecycleScope.launch {
                            login()
                        }
                    }
                    .onDenied {
                        TipDialog.show(this@MainActivity,"缺少必要权限",TipDialog.TYPE.ERROR)
                        RtmManager.instance.release()
                        RtcManager.instance.release()
                        exitProcess(0)
                        finish()
                    }
                    .start()

    }

    private suspend fun MainActivity.login() {
        val result = loginRtm()
        loginTipDialog?.doDismiss()
        if (result) {
            isLoginSuccess = true
            TipDialog.show(this@MainActivity, "登录成功！", TipDialog.TYPE.SUCCESS)
        } else {
            isLoginSuccess = false
            TipDialog.show(this@MainActivity, "登录失败！", TipDialog.TYPE.SUCCESS)
        }
    }


    private fun showLoginTip(){
        loginTipDialog = MessageDialog.build(this@MainActivity)
                .setTitle("提示")
                .setMessage("正在登录....")
                .setOnCancelButtonClickListener { _, _ ->
                    true
                }
                .setCancelable(false)
        loginTipDialog?.show()
    }

    private suspend fun loginRtm() = suspendCoroutine<Boolean> {
        showLoginTip()
        RtmManager.instance.getRtmClient()?.login("",callApp.userId,object :ResultCallback<Void>{
            override fun onSuccess(var1: Void?) {
                it.resume(true)
            }

            override fun onFailure(var1: ErrorInfo?) {
                it.resume(false)
            }

        })
    }

    private fun showExitDialog() {
        MessageDialog.show(this, "提示", "确定要退出吗？", "确定")
                .setCancelButton("取消") { baseDialog: BaseDialog, v: View? ->
                    baseDialog.doDismiss()
                    true
                }.onOkButtonClickListener = OnDialogButtonClickListener { baseDialog: BaseDialog, v: View? ->
            baseDialog.doDismiss()
            RtcManager.instance.release()
            exitProcess(0)
            finish()
            true
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            showExitDialog()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    fun single(view: View?) {
        if (isLoginSuccess) {
            val intent = Intent(this@MainActivity, CallActivity::class.java)
            startActivity(intent)
        }else{
            lifecycleScope.launch {
                login()
            }
        }
    }

    fun multiple(view: View?) {
        if (isLoginSuccess) {
            val intent = Intent(this@MainActivity, MultiCallActivity::class.java)
            startActivity(intent)
        }else{
            lifecycleScope.launch {
                login()
            }
        }
    }


    override fun onRemoteInvitationReceived(remoteInvitation: RemoteInvitation) {
        super.onRemoteInvitationReceived(remoteInvitation)
        runOnUiThread {
            if (EasyFloat.appFloatIsShow()){
                return@runOnUiThread
            }
                val jsonObject = JSONObject(remoteInvitation.content)
                val isConference = jsonObject["Conference"]

                Intent().apply {
                    putExtra("RecCall",true)
                    setClass(this@MainActivity,if (isConference==1||isConference==true){
                        MultiCallActivity::class.java
                    }else{
                        VideoActivity::class.java
                    })
                    startActivity(this)
                }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
    }
}